ludwig/modules/serialization.py

```diff
@@ -0,0 +1,149 @@
+import logging
+from functools import lru_cache
+from typing import Any, Dict
+
+import h5py
+import numpy as np
+
+from ludwig.globals import LUDWIG_VERSION
+from ludwig.modules.ludwig_module import LudwigModule, LudwigModuleState, module_registry
+from ludwig.utils.fs_utils import download_h5, upload_h5
+
+logger = logging.getLogger(__name__)
+
+
+class NumpyEncoder(json.JSONEncoder):
+    """Python json library does not support serialization of numpy types.
+
+    This custom encoder converts numpy scalars to python types
+    """
+
+    def default(self, obj):
+        if isinstance(
+            obj,
+            (
+                np.int_,
+                np.intc,
+                np.intp,
+                np.int8,
+                np.int16,
+                np.int32,
+                np.int64,
+                np.uint8,
+                np.uint16,
+                np.uint32,
+                np.uint64,
+            ),
+        ):
+            return int(obj)
+        elif isinstance(obj, (np.float_, np.float16, np.float32, np.float64)):
+            return float(obj)
+        elif isinstance(obj, (np.complex_, np.complex64, np.complex128)):
+            return {"real": obj.real, "imag": obj.imag}
+        elif isinstance(obj, (np.ndarray,)):
+            return obj.tolist()
+        elif isinstance(obj, (np.bool_)):
+            return bool(obj)
+        elif isinstance(obj, (np.void)):
+            return None
+        return json.JSONEncoder.default(self, obj)
+
+
+def _save_state_to_group(state: LudwigModuleState, group: h5py.Group):
+    """Save LudwigModuleState to HDF5 group, recursively creating groups for any child modules."""
+    group.attrs["type"] = state.type
+    group.attrs["ludwig_version"] = state.ludwig_version
+    group.attrs["config"] = json.dumps(state.config, cls=NumpyEncoder, sort_keys=True)
+    group.attrs["metadata"] = json.dumps(state.metadata, cls=NumpyEncoder, sort_keys=True)
+    for k, w in state.saved_weights.items():
+        group.create_dataset(k, data=w)
+    for k, child in state.children.items():
+        child_group = group.create_group(k)
+        _save_state_to_group(child, child_group)
+
+
+def save_state_to_file(state: LudwigModuleState, url_or_path: str):
+    """Serializes LudwigModuleState to file."""
+    with upload_h5(url_or_path) as f:
+        # with h5py.File(f, "w") as f:
+        _save_state_to_group(state, f)
+
+
+def _load_state_from_group(group) -> LudwigModuleState:
+    """Restore LudwigModuleState from HDF5 group."""
+    return LudwigModuleState(
+        type=group.attrs["type"],
+        ludwig_version=group.attrs["ludwig_version"],
+        config=json.loads(group.attrs["config"]),
+        metadata=json.loads(group.attrs["metadata"]),
+        saved_weights={k: v[()] for k, v in group.items() if isinstance(v, h5py.Dataset)},
+        children={k: _load_state_from_group(v) for k, v in group.items() if isinstance(v, h5py.Group)},
+    )
+
+
+@lru_cache(maxsize=3)
```

Comment:
Curious -- how did you determine 3 as the maxsize to use?

Reply:
No particular reason.